proposals/0166-dynamic-stack-frames.md

```diff
@@ -0,0 +1,173 @@
+
+The SVM allocates a fixed amount of memory to hold a program’s stack. Within 
+the stack region, the virtual machine reserves 4096 bytes of stack space for 
+each function frame. This is simultaneously limiting for functions that 
+require more space, and wasteful for functions that require less space.
+
+For well optimized programs that don’t allocate large amounts of stack, the 
+virtual machine currently still reserves 4096 bytes of stack for each 
+function call, leading to suboptimal memory usage, which may cause 
+unnecessary page faults.
+
+On the other hand, some programs are known to create large function frames - 
+this seems common with programs that serialize a lot of data - and they have 
+to jump through hoops to avoid overflowing the stack. The virtual machine 
+detects when a stack overflow occurs, and it does so by implementing a stack 
+frame gaps system whereby it inserts a virtual sentinel frame following a 
+valid function frame. If the sentinel frame is accessed, the executing program 
+is aborted. This system is fragile and is incompatible with direct mapping - 
+a feature we expect to enable soon. 
+
+The changes proposed in this document would allow us to optimize stack memory 
+usage and remove the fragile stack frame gaps system. Note that we do not 
+propose to remove the existing maximum stack space limit: stack space stays 
+unchanged, what changes is how it is partitioned internally.
+
+## Alternatives Considered
+
+To cope with the SBF limitation of 4096 bytes for the frame size, we could 
+have increased such a number. Even though this would solve the original 
+problem, it would supply an unnecessary amount of memory to functions even 
+when they do not need them. In addition, such a solution would increase 
+pressure on the total memory available for the call stack. Either we would 
+need to increase the total allocation for the virtual machine or decrease the 
+maximum call depth.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+Bringing dynamic stack frames to the Solana Bytecode Format and its 
+corresponding virtual machine entails changes in several aspects of the 
+execution environment.
+
+### SBF architecture modifications
+
+
+We will introduce a new register R11 in the virtual machine, which is going 
+to hold the stack pointer. The program must only write to such a register and 
+modify it through the `add64 reg, imm` (op code `0x07`) instruction. The 
+verifier must enforce these constraints on deployed programs. For further 
+information about the  changes in the ISA, refer to the 
+[SPF spec document](https://github.com/solana-labs/rbpf/blob/main/doc/bytecode.md).
+
+The R11 register must work in tandem with the R10 (frame pointer) register. 
+The former is write-only to the program, and the latter is read-only to the 
+program, forming a common design pattern in hardware engineering. More 
+details of this usage are in the following section.
+
+### Changes in the execution environment
```

Comment:
I'd add the following section.

### Stack alignment

We want to enforce that the stack pointer remains aligned, therefore we impose that R11 can only be incremented or decremented by a multiple of the chosen alignment. Superficially, larger alignment might seem a bit wasteful, but we believe that enforcing a sufficiently large alignment will lead to innovation in interpreters and JITs, ultimately leading to much better performance and thus lower costs to the users.

Based on the current AVX512 instructions available on Intel and AMD processors, we set the stack alignment to 64 bytes.

Even if current interpreters do not take advantage of any of these vectorized instructions, we believe that future generation interpreters might be able to vectorize SBF programs, for example to speed up common operations such as copying or comparing public keys or signatures. Unaligned stack would prohibit this kind of innovation.

Reply:
Added